pkg/controllers/node/termination/terminator/terminator.go

```diff
@@ -46,18 +45,18 @@ func NewTerminator(clk clock.Clock, kubeClient client.Client, eq *Queue) *Termin
 	}
 }
 
-// Taint idempotently adds the karpenter.sh/disruption taint to a node with a NodeClaim
-func (t *Terminator) Taint(ctx context.Context, node *v1.Node) error {
+// Taint idempotently adds an arbitrary taint to a node with a NodeClaim
+func (t *Terminator) Taint(ctx context.Context, node *v1.Node, taint v1.Taint) error {
 	stored := node.DeepCopy()
-	// If the taint already has the karpenter.sh/disruption=disrupting:NoSchedule taint, do nothing.
+	// If the node already has the correct taint (key, value, and effect), do nothing.
 	if _, ok := lo.Find(node.Spec.Taints, func(t v1.Taint) bool {
-		return v1beta1.IsDisruptingTaint(t)
+		return t.MatchTaint(&taint) && t.Value == taint.Value && t.Effect == taint.Effect
 	}); !ok {
-		// If the taint key exists (but with a different value or effect), remove it.
+		// Otherwise, if the taint key exists (but with a different value or effect), remove it.
 		node.Spec.Taints = lo.Reject(node.Spec.Taints, func(t v1.Taint, _ int) bool {
-			return t.Key == v1beta1.DisruptionTaintKey
+			return t.Key == taint.Key
 		})
-		node.Spec.Taints = append(node.Spec.Taints, v1beta1.DisruptionNoScheduleTaint)
```

Comment:
I rewrote this helper to handle arbitrary Taints, so it can be used for both the karpenter disrupting taint and the non-graceful shutdown taint.